Removing card ui

DDKatch opened the pull request:
How it works
![out](https://user-images.githubusercontent.com/14095146/50351284-c14cbe00-0552-11e9-8d64-afcc650ba151.gif)

DDKatch commented:
![](https://github.trello.services/images/mini-trello-icon.png) [I as user want to delete TODO card](https://trello.com/c/ZZt7Tspg/22-i-as-user-want-to-delete-todo-card)

AleksSenkou commented on client/src/components/TodoCardList/Item/index.js:
```diff
@@ -1,17 +1,29 @@
-import { Card, CardBody, Input } from 'mdbreact'
+import { Card, CardBody } from 'mdbreact'
+
+import styles from 'src/components/TodoCardList/Item/styles.css'
```
@DDKatch can't we do 
`import styles from 'styles.css'`
or
`import styles from './styles.css'`
?

AleksSenkou commented on client/src/components/TodoCardList/Item/styles.css:
```diff
@@ -0,0 +1,7 @@
+a{
```
`a {`

`a:hover {`

AleksSenkou commented on client/src/components/TodoCardList/Item/index.js:
```diff
@@ -1,17 +1,29 @@
+          <span className="col-sm-8 text-left">
+            {props.description}
+          </span>
+          <a className="col-sm-2" href="http://localhost:3000/" styles={styles}>
```
@datarockets/cuties thoughts about having a file called `server.js` in the root of react app in order to keep all routes there?

So, we'll have:
```js
import server from 'src/server'

...

<a className="col-sm-2" href=`${server.baseUrl}` styles={styles}>
```

AleksSenkou reviewed (dismissed):
@DDKatch I can't find axios request to the server and predefined endpoint (blueprint doc with skipped test)

Are you going to add them in this PR?

DDKatch committed:
Add button for card deletion

DDKatch committed:
Add ajax query for deleting todo cards

DDKatch commented:
@datarockets/cuties as you can see I've added some infra tips for client such as `server.baseUrl` and refactored requests a little. What do you think about scale this changes for other _request.js_ files ?

DDKatch commented:
@AleksSenkou let's mount that changes for docs and tests will be shipped with backend features, because it's obviously will crash during hypothetical ci jobs. Thoughts?

AleksSenkou commented:
@DDKatch it won't be crashed during CI cause we can skip tests before server changes this way https://github.com/datarockets/loopback-to-do/pull/22/files#r242062175

DDKatch commented:
@AleksSenkou got it, going to add tests here

DDKatch committed:
Add doc for endpoint DELETE /api/v1/cards/{id}

DDKatch commented:
@AleksSenkou I've added docs, but test fails as I expected. What should I do in order to skip it?
<img width="464" alt="screen shot 2018-12-22 at 4 09 49 pm" src="https://user-images.githubusercontent.com/14095146/50374732-2ca29880-0604-11e9-8b7e-f796daebc47b.png">

AleksSenkou commented:
@DDKatch https://github.com/datarockets/loopback-to-do/pull/25#issuecomment-449438276

Please, check it and ask what is unclear

DDKatch commented on client/src/containers/App/index.js:
```diff
@@ -18,12 +18,17 @@ class AppContainer extends Component {
   }
+
+  deleteTodoCard = (cardId) => {
+    this.setState(state => ({ todoCards: state.todoCards.filter(el => el.id !== cardId) }))
```
@AleksSenkou i think it's not so long. Any way which one do you prefer:
1.```
 this.setState(state =>
      ({ todoCards: state.todoCards.filter(el => el.id !== cardId) }))

```
2.```
    this.setState(
      state => ({ todoCards: state.todoCards.filter(el => el.id !== cardId) }),
    )
```

AleksSenkou replied:
@DDKatch hm, what do you think about this way:
```js
deleteTodoCard = cardId =>
  this.setState({todoCards} => { todoCards: this._filteredCards(todoCards, cardId) })

_filteredCards = (cards, cardId) =>
  cards.filter(card => card.id !== cardId)
```

method with prefix `_` used for showing that it's private // @datarockets/cuties

DDKatch replied:
@AleksSenkou I don't like it, even this line even longer than mine :wink: . Even in naming I prefer my variant, because `_filteredCards` is a really extensive name.

AleksSenkou replied:
@DDKatch we can go without line splitting then 👍 

So we'll split that when we need to change logic of filtering cards

DDKatch commented on client/src/containers/App/index.js:
```diff
@@ -18,12 +18,17 @@ class AppContainer extends Component {
     this.setState(state => ({ todoCards: [...state.todoCards, card] }))
   }
+
+  deleteTodoCard = (cardId) => {
```
@AleksSenkou linter doesn't support it in this case 🤔

AleksSenkou replied:
@DDKatch can you give an error example and other options?

AleksSenkou replied:
@DDKatch I guess it is `Expected parentheses around arrow function argument having a body with curly braces` that seems ok because the body is rounded with `{}`

AleksSenkou replied:
I'm going to fix it in https://trello.com/c/xkkmIu2M/40-when-i-develop-the-app-id-like-not-to-see-linter-errors

DDKatch replied:
@AleksSenkou let's merge this pr after yours because of that

DDKatch replied:
@AleksSenkou I've rebased it on the new master

AleksSenkou commented on client/src/components/TodoCardList/Item/DeletionButton/index.js:
```diff
@@ -0,0 +1,25 @@
+}
+
+DeletionButton.defaultProps = {
+  onClick: () => alert('DeletionButton onClick Handler'),
```
@datarockets/cuties there is mock for that in the community:

```js
import { noop } from underscore

defaultProps = {
  onClick: noop
}
```

https://underscorejs.org/#noop

AleksSenkou commented on client/src/containers/App/index.js:
```diff
@@ -18,12 +18,17 @@ class AppContainer extends Component {
+
+  deleteTodoCard = (cardId) => {
+    this.setState(state => ({ todoCards: state.todoCards.filter(el => el.id !== cardId) }))
+  }
```
@DDKatch https://trello.com/c/xkkmIu2M/40-when-i-develop-the-app-id-like-not-to-see-linter-errors is merged (nothing is changed in linter), so you can go with:

```js
deleteTodoCard = cardId =>
  this.setState(state => ({ todoCards: state.todoCards.filter(el => el.id !== cardId) }))
```
instead of:
```js
deleteTodoCard = (cardId) => {
  this.setState(state => ({ todoCards: state.todoCards.filter(el => el.id !== cardId) }))
}
```

AleksSenkou commented on client/src/containers/TodoCardList/Item/index.js:
```diff
@@ -13,12 +13,21 @@ class TodoCardContainer extends Component {
+    api.cards.delete(
+      { id: this.props.id },
+      () => { this.props.onDelete(this.props.id) },
+      (error) => { alert(error) },
```
```js
() => this.props.onDelete(this.props.id),
error => alert(error),
```
can be used instead, thoughts?

AleksSenkou commented on client/src/server.js:
```diff
@@ -0,0 +1,9 @@
+  cards: () => `${routes.baseUrl}/cards`,
+  card: id => `${routes.baseUrl}/card/${id}`,
+  cardCompletion: id => `${routes.card(id)}/completion`,
+}
```
❤️

AleksSenkou commented on client/src/server.js:
```diff
@@ -0,0 +1,9 @@
+const routes = {
+  baseUrl: 'http://localhost:5000/api/v1',
+
+  cards: () => `${routes.baseUrl}/cards`,
```
@DDKatch can't we go with just `baseUrl`?

AleksSenkou commented on server/package.json:
```diff
@@ -12,16 +12,12 @@
-
     "build": "babel src -d build",
     "start": "node build",
-
```
@DDKatch I was thinking about living these new lines for better groups separation, thoughts?

AleksSenkou commented on server/package.json:
```diff
@@ -36,15 +32,15 @@
     "pg-hstore": "^2.3.2",
     "sequelize": "^4.41.2",
-    "sequelize-cli": "^5.3.0"
+    "sequelize-cli": "^5.4.0"
```
Great!

AleksSenkou commented on client/src/containers/TodoCardList/Item/index.js:
```diff
@@ -41,14 +54,15 @@ class TodoCardContainer extends Component {
   }
```
Seems like indentantion is crazy here 😄 
```
}
}
```

AleksSenkou commented on server/spec/unit/controllers/Cards/deletion.hook.js:
```diff
@@ -0,0 +1,9 @@
+import { before } from 'hooks'
```
Guess file should be under `system` domain, not `unit` 😉

AleksSenkou replied:
Anyway that's not needed, because we already have backend for that

BTW, we need to prepare testing database before several steps, we can do it via PP if you want (separate infra task can be created for that)

DDKatch replied:
@AleksSenkou ok, let's do it tomorrow

DDKatch replied:
@AleksSenkou opened task for that https://trello.com/c/R2QnJBZa/45-when-i-develop-app-id-like-to-write-different-sorts-of-test

AleksSenkou commented on server/api-docs.apib:
```diff
@@ -73,6 +73,41 @@ Creates Card
+
++ Response 204
+
+## Card not found [/api/v1/cards/{id}]
```
![image](https://user-images.githubusercontent.com/3635774/50490203-d7f48880-0a1c-11e9-9948-5f19ce93cd96.png)

DDKatch commented on client/src/components/TodoCardList/Item/DeletionButton/index.js:
```diff
@@ -0,0 +1,25 @@
+)
+
+DeletionButton.propTypes = {
+  onClick: PropTypes.func,
```
<img width="649" alt="screen shot 2018-12-27 at 9 26 59 pm" src="https://user-images.githubusercontent.com/14095146/50490625-1ee37d80-0a1f-11e9-8cef-486f27cc4f95.png">

AleksSenkou replied:
@DDKatch I've tried to simulate stack trace, but didn't find a problem, you may want to ask in #dev chat about ideas and pops of it

DDKatch replied:
@AleksSenkou got it, will post in #dev

DDKatch replied:
@AleksSenkou Opened card for that https://trello.com/c/kI3i1Xcr/44-when-i-develop-app-id-like-to-code-properly-without-some-not-bad-but-anyway-hacks

AleksSenkou commented on server/api-docs.apib:
````diff
@@ -73,6 +73,41 @@ Creates Card
+        error: "Bad Request",
+        message: "Card not found"
+      }
+      ```
````
Yeah, apiblueprint is a little bit unclear base we can't separate blocks like we would like to. Super strict rules for that. Here is overwritten docs for provided above (I've rewritten this part with a look on [docs](https://github.com/apiaryio/api-blueprint/edit/master/examples/Real%20World%20API.md)):

```
## Card [/api/v1/cards/{id}]

- Parameters
    + id: 1 (required)

### DELETE

- Request (application/json)
- Response 204

- Request (application/json)

    When record not found

    - Parameters
        + id: 1000 (required)

- Response 400

    - Attributes (object)
        + error (string) - Error type
        + message (string) - Error message

    - Body

        ```
        {
          error: "Bad Request",
          message: "Card not found"
        }
        ```
```

DDKatch replied:
@AleksSenkou 
it fails everytime:
```

- Request (application/json)
    - Parameters
        + id: 1000 (required)

- Response 400

    - Attributes (object)
        + error (string) - Error type
        + message (string) - Error message

    - Body

        ```
        {
          error: "Bad Request",
          message: "Card not found"
        }
        ```

```

DDKatch replied:
Fixed during PP

AleksSenkou replied:
`(application/json; charset=utf-8)`

AleksSenkou commented on client/src/server.js:
```diff
@@ -0,0 +1,9 @@
+  baseUrl: 'http://localhost:5000/api/v1',
+
+  cards: () => `${routes.baseUrl}/cards`,
+  card: id => `${routes.baseUrl}/card/${id}`,
```
`cards` instead of `card`, locally fails, wrong url

AleksSenkou commented on client/src/containers/TodoCardList/Item/requests.js:
```diff
@@ -1,26 +1,32 @@
+    delete: (card, responseHandler, errorHandler) => request(
+      axios.delete(routes.card(card.id)),
+      errorHandler,
+      responseHandler,
```
responseHandler and errorHandler are missed (wrong positions)

DDKatch replied:
@AleksSenkou I've done this purposely in order to deliver working code, but forgot about that after squashing commits. Will pay more attention in the future cases.

DDKatch commented on client/src/components/TodoCardList/Item/index.js:
```diff
@@ -1,17 +1,20 @@
+      <CardBody>
+        <div className="row d-flex align-items-center">
+          <CompletionCheckbox onChange={props.onChange} checked={props.completed} />
+          <Description description={props.description} />
```
@AleksSenkou let's go with `text`

nikitasakov commented:
@DDKatch Something strange on this branch,( https://www.youtube.com/watch?v=WwTPyrtl0DE&feature=youtu.be ) Have you got the same problem?

AleksSenkou commented:
@nikitasakov @DDKatch seems like this branch have to be rebased with master before any check

AleksSenkou commented on client/src/containers/TodoCardList/Item/index.js:
```diff
@@ -13,18 +13,27 @@ class TodoCardContainer extends Component {
+  deleteCard = () => {
+    api.cards.delete(
+      { id: this.props.id },
+      () => { this.props.onDelete(this.props.id) },
```
thoughts about not using `{ .. }` if possible

```js
() => this.props.onDelete(this.props.id),
```
?

DDKatch replied:
@AleksSenkou, ups I've pushed the horses. I don't see much difference and our linter says my code is ok, let's change linter rules next time, because it corrects my code automatically

AleksSenkou reviewed (approved):
LGTM with comment

DDKatch commented:
> @DDKatch Something strange on this branch,( https://www.youtube.com/watch?v=WwTPyrtl0DE&feature=youtu.be ) Have you got the same problem?

I've already fixed 😉
